app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -137,6 +136,183 @@ object DialogUtils {
+            sheet.dismiss()
+        }
+        /**
+         * Function that updates all the views in the gradient.
+         * Takes in a color state data class to update views.
+         */
+        fun forceViewUpdates(colorState: ColorUtils.ColorState) {
+            val displaySat = (colorState.saturation * 100).toInt()
+            val displayVal = (colorState.value * 100).toInt()
+            hexText.forceText(colorState.hexCode)
+            hueText.forceText(colorState.hue.toString())
+            satText.forceText(displaySat.toString())
+            valText.forceText(displayVal.toString())
+            redText.forceText(colorState.red.toString())
+            greenText.forceText(colorState.green.toString())
+            blueText.forceText(colorState.blue.toString())
+            huePicker.setHue(colorState.hue)
+            gradientPicker.setHue(colorState.hue)
+            gradientPicker.setSaturationAndValue(colorState.saturation, colorState.value)
+            topColorSquare.setBackgroundColor(parseColor(colorState.hexCode))
+        }
+
+        /**
+         * Clears focus on whatever view the user has currently selected.
+         * Currently only called when the user interacts with the Gradient or Hue Slider.
+         */
+        fun clearFocus() {
+            hexText.clearFocus()
+            hueText.clearFocus()
+            satText.clearFocus()
+            valText.clearFocus()
+            redText.clearFocus()
+            greenText.clearFocus()
+            blueText.clearFocus()
+        }
+        val colorStates = ColorUtils.calculateColorWithHSV(previousColor)
+        forceViewUpdates(colorStates)
+        bottomColorSquare.setBackgroundColor(parseColor(colorStates.hexCode))
+        huePicker?.onHueChanged = {
+            clearFocus()
+            val hsv = FloatArray(3)
+            hsv[0] = it.toFloat()
+            hsv[1] = (satText.text.toString().toFloat() / 100)
+            hsv[2] = (valText.text.toString().toFloat() / 100)
+            val calculatedColorState = ColorUtils.calculateColorWithHSV(hsv)
+            gradientPicker.setHue(it)
+            forceViewUpdates(calculatedColorState)
+        }
+        gradientPicker?.onSatValChanged = { saturation, value ->
+            clearFocus()
+            val hsv = FloatArray(3)
+            hsv[0] = (hueText.text.toString().toFloat())
+            hsv[1] = saturation
+            hsv[2] = 1f - value
+            val calculatedColorState = ColorUtils.calculateColorWithHSV(hsv)
+            forceViewUpdates(calculatedColorState)
+        }
+        hexText.addTextChangedListener {
+            if (ColorUtils.isHexadecimal(it.toString())) {
+                val color = parseColor("#${it.toString()}")
+                val rgb = IntArray(3)
+                rgb[0] = Color.red(color)
+                rgb[1] = Color.green(color)
+                rgb[2] = Color.blue(color)
+                val calculatedColorState = ColorUtils.calculateColorWithRGB(rgb)
```

Comment:
why not just pass floatArrayOf(Color.red(color),Color.green(color),Color.blue(color)).

Reply:
That would be better in this case! Implementing this.